apps/emails-and-messages/src/modules/sendgrid/ui/api-connection-section.tsx

```diff
@@ -0,0 +1,92 @@
+import { SendgridConfiguration } from "../configuration/sendgrid-config-schema";
+import { BoxWithBorder } from "../../../components/box-with-border";
+import { Box, Button, Text } from "@saleor/macaw-ui/next";
+import { defaultPadding } from "../../../components/ui-defaults";
+import { useDashboardNotification } from "@saleor/apps-shared";
+import { trpcClient } from "../../trpc/trpc-client";
+import {
+  SendgridUpdateApiConnection,
+  sendgridUpdateApiConnectionSchema,
+} from "../configuration/sendgrid-config-input-schema";
+import { useForm } from "react-hook-form";
+import { BoxFooter } from "../../../components/box-footer";
+import { SectionWithDescription } from "../../../components/section-with-description";
+import { Input } from "../../../components/react-hook-form-macaw/Input";
+import { zodResolver } from "@hookform/resolvers/zod";
+
+interface ApiConnectionSectionProps {
+  configuration: SendgridConfiguration;
+}
+
+export const ApiConnectionSection = ({ configuration }: ApiConnectionSectionProps) => {
+  const { notifySuccess, notifyError } = useDashboardNotification();
+
+  const { handleSubmit, control, setError, register } = useForm<SendgridUpdateApiConnection>({
+    defaultValues: {
+      id: configuration.id,
+      apiKey: configuration.apiKey,
+      sandboxMode: configuration.sandboxMode,
+    },
+    resolver: zodResolver(sendgridUpdateApiConnectionSchema),
+  });
+
+  const trpcContext = trpcClient.useContext();
+  const { mutate } = trpcClient.sendgridConfiguration.updateApiConnection.useMutation({
+    onSuccess: async (data, variables) => {
+      notifySuccess("Configuration saved");
+      trpcContext.sendgridConfiguration.invalidate();
+    },
+    onError(error) {
+      let isFieldErrorSet = false;
+      const fieldErrors = error.data?.zodError?.fieldErrors || {};
+      for (const fieldName in fieldErrors) {
+        for (const message of fieldErrors[fieldName] || []) {
+          isFieldErrorSet = true;
+          setError(fieldName as keyof SendgridUpdateApiConnection, {
+            type: "manual",
+            message,
+          });
+        }
+      }
```

Comment:
Id extract this

Reply:
Had to play with generics, but got this :muscle: